File: src/main/java/net/ndrei/teslapoweredthingies/gui/GeneratorBurnPiece.java
package net.ndrei.teslapoweredthingies.gui;

import com.google.common.collect.Lists;
import com.mojang.realmsclient.gui.ChatFormatting;
import net.darkhax.tesla.lib.TeslaUtils;
import net.ndrei.teslacorelib.gui.BasicContainerGuiPiece;
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer;
import net.ndrei.teslacorelib.tileentities.ElectricGenerator;
import net.ndrei.teslapoweredthingies.TeslaThingiesMod;

import java.util.List;

/**
 * Created by CF on 2017-01-07.
 */
public class GeneratorBurnPiece extends BasicContainerGuiPiece {
    private ElectricGenerator te;

    public GeneratorBurnPiece(int left, int top, ElectricGenerator te) {
        super(left, top, 14, 14);

        this.te = te;
    }

    @Override
    public void drawBackgroundLayer(BasicTeslaGuiContainer container, int guiX, int guiY, float partialTicks, int mouseX, int mouseY) {
        container.mc.getTextureManager().bindTexture(TeslaThingiesMod.MACHINES_TEXTURES);

        container.drawTexturedRect(this.getLeft(), this.getTop(), 44, 27, 14, 14);
        long generated = this.te.getGeneratedPowerCapacity();
        long stored = this.te.getGeneratedPowerStored();
        int percent = Math.round(14 * Math.min(Math.max((float)(generated - stored) / (float)generated, 0), 1));
        if (percent > 0) {
            container.drawTexturedRect(this.getLeft(), this.getTop() + percent, 8, 27 + percent, 14, 14 - percent);
        }
    }

    @Override
    public void drawForegroundTopLayer(BasicTeslaGuiContainer container, int guiX, int guiY, int mouseX, int mouseY) {
        if (!this.isInside(container, mouseX, mouseY)) {
            return;
        }

        List<String> lines = GeneratorBurnPiece.getTooltipLines(this.te);
        if ((lines != null) && (lines.size() > 0)) {
            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY);
        }
    }

    public static List<String> getTooltipLines(ElectricGenerator entity) {
        List<String> lines = Lists.newArrayList();
        long generated = entity.getGeneratedPowerCapacity();
        if ((generated > 0) && (entity.getGeneratedPowerStored() > 0)) {
            lines.add(ChatFormatting.GRAY + "Total for fuel: "
                    + ChatFormatting.AQUA + TeslaUtils.getDisplayableTeslaCount(generated));
            lines.add(ChatFormatting.GRAY + "Generating "
                    + ChatFormatting.AQUA + TeslaUtils.getDisplayableTeslaCount(entity.getGeneratedPowerReleaseRate())
                    + ChatFormatting.GRAY + " / tick");

            double ticks = (double)entity.getGeneratedPowerStored() / (double)entity.getGeneratedPowerReleaseRate() / 20.0;
            lines.add(ChatFormatting.GRAY + "~ " + String.format("%.2f", ticks) + "s remaining");
        }
        return lines;
    }
}

File: src/main/kotlin/net/ndrei/teslapoweredthingies/gui/GeneratorBurnPiece.kt
package net.ndrei.teslapoweredthingies.gui

import com.google.common.collect.Lists
import com.mojang.realmsclient.gui.ChatFormatting
import net.darkhax.tesla.lib.TeslaUtils
import net.ndrei.teslacorelib.gui.BasicContainerGuiPiece
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer
import net.ndrei.teslacorelib.tileentities.ElectricGenerator
import net.ndrei.teslapoweredthingies.client.Textures

/**
 * Created by CF on 2017-06-30.
 */
class GeneratorBurnPiece(left: Int, top: Int, private val te: ElectricGenerator)
    : BasicContainerGuiPiece(left, top, 14, 14) {

    override fun drawBackgroundLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
        // container.mc.textureManager.bindTexture(TeslaThingiesMod.MACHINES_TEXTURES)
        Textures.MACHINES_TEXTURES.bind(container)

        container.drawTexturedRect(this.left, this.top, 44, 27, 14, 14)
        val generated = this.te.generatedPowerCapacity
        val stored = this.te.generatedPowerStored
        val percent = Math.round(14 * Math.min(Math.max((generated - stored).toFloat() / generated.toFloat(), 0f), 1f))
        if (percent > 0) {
            container.drawTexturedRect(this.left, this.top + percent, 8, 27 + percent, 14, 14 - percent)
        }
    }

    override fun drawForegroundTopLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, mouseX: Int, mouseY: Int) {
        if (!this.isInside(container, mouseX, mouseY)) {
            return
        }

        val lines = GeneratorBurnPiece.getTooltipLines(this.te)
        if (lines != null && lines.size > 0) {
            container.drawTooltip(lines, mouseX - guiX, mouseY - guiY)
        }
    }

    companion object {

        fun getTooltipLines(entity: ElectricGenerator): List<String>? {
            val lines = Lists.newArrayList<String>()
            val generated = entity.generatedPowerCapacity
            if (generated > 0 && entity.generatedPowerStored > 0) {
                lines.add(ChatFormatting.GRAY.toString() + "Total for fuel: "
                        + ChatFormatting.AQUA + TeslaUtils.getDisplayableTeslaCount(generated))
                lines.add(ChatFormatting.GRAY.toString() + "Generating "
                        + ChatFormatting.AQUA + TeslaUtils.getDisplayableTeslaCount(entity.generatedPowerReleaseRate)
                        + ChatFormatting.GRAY + " / tick")

                val ticks = entity.generatedPowerStored.toDouble() / entity.generatedPowerReleaseRate.toDouble() / 20.0
                lines.add(ChatFormatting.GRAY.toString() + "~ " + java.lang.String.format("%.2f", ticks) + "s remaining")
            }
            return lines
        }
    }
}